### ISSUE TITLE
By layer uncertainty tweaks

### ISSUE THREAD

**Issue by Eupham** [Eupham/KernelDev#129]
Below is a crisp, file-by-file plan to implement deep supervision at every layer for both tasks with a shared head and per-(layer,task) uncertainties. It also extends the loop to compute contrastive span-selection losses at intermediate layers. I’ll keep it tight: bullet points, code, and math.

Overview — What you do now vs what we want

Now

Intermediate (layerwise) losses only for teacher_forcing and only every _N_th layer.

Single log_sigma per layer (shared across tasks) or missing for some layers.

Cocktail party (cp) has only final CE; no intermediate losses.

Reported loss sometimes mixes raw CE with uncertainty weighting.

Wanted

All layers accounted for (every block contributes).

Both tasks accounted for at every supervised layer.

Per-(layer, task) uncertainty parameters 
𝑠
ℓ
,
𝑡
=
log
⁡
𝜎
ℓ
,
𝑡
s
ℓ,t
	​

=logσ
ℓ,t
	​

.

Shared head across layers (no head per layer). Use scalar conditioning to tell the head “which layer” and “which task”.

Raw CE and uncertainty contributions reported separately.

Math (final objective)

For layer 
ℓ
∈
{
1..
𝐿
}
ℓ∈{1..L}, task 
𝑡
∈
{
tf
,
cp
}
t∈{tf,cp}:

𝐿
ℓ
,
𝑡
unc
=
1
2
 
𝑒
−
2
𝑠
ℓ
,
𝑡
 
𝐿
ℓ
,
𝑡
raw
+
𝑠
ℓ
,
𝑡
L
ℓ,t
unc
	​

=
2
1
	​

e
−2s
ℓ,t
	​

L
ℓ,t
raw
	​

+s
ℓ,t
	​

	​

𝐿
total
=
𝐿
final,tf
+
𝐿
final,cp
+
∑
ℓ
=
1
𝐿
∑
𝑡
∈
{
tf
,
cp
}
𝜆
𝑡
 
𝐿
ℓ
,
𝑡
unc
+
𝜆
kl
∑
ℓ
,
𝑡
𝑠
ℓ
,
𝑡
2
L
total
	​

=L
final,tf
	​

+L
final,cp
	​

+
ℓ=1
∑
L
	​

t∈{tf,cp}
∑
	​

λ
t
	​

L
ℓ,t
unc
	​

+λ
kl
	​

ℓ,t
∑
	​

s
ℓ,t
2
	​

	​

Report (metrics): headline losses = 
𝐿
final,tf
L
final,tf
	​

 and 
𝐿
final,cp
L
final,cp
	​

 (raw CE).
Log layer CE means and uncertainty deltas separately.

File-by-file edits
1) configs/*.yaml (or wherever)

supervise_layers: "all" # all blocks contribute

share_heads: true

conditioning: "film" # or "concat2" (see below)

lambda_layers: { tf: 1.0, cp: 1.0 }

lambda_kl: 1e-3

2) model.py
2.1 Shared heads (no per-layer heads)

Teacher forcing (LM) shared head (one matrix for all layers):

self.shared_tf_head = nn.Linear(self.dim + (2 if cfg.conditioning=="concat2" else 0),
                                self.vocab_size, bias=False)
# If you already tie to embeddings, you can project to dim first and then use tied weights.

Cocktail party uses span dot products; no per-layer “head” is needed. (Optional: a single shared projection for spans/context if you want a learned metric.)

2.2 Per-(layer, task) uncertainties

Create learnable log-σ for each (layer, task):

task_names = ["teacher_forcing","cocktail_party"]

self.layer_log_sigmas = nn.ModuleList([
    nn.ParameterDict({
        t: nn.Parameter(torch.normal(mean=0.0, std=0.02, size=(1,)))  # s_{ℓ,t}
        for t in task_names
    }) for _ in range(self.n_layers)
])

2.3 Scalar conditioning (lets the shared head “know” layer & task)

Two options; pick one via cfg.conditioning:

A) FiLM-style (no shape change):

Learn per-layer and per-task scalars 
𝛼
ℓ
,
𝛽
ℓ
,
𝛼
𝑡
,
𝛽
𝑡
α
ℓ
	​

,β
ℓ
	​

,α
t
	​

,β
t
	​

Apply: 
ℎ
′
=
(
1
+
𝛼
ℓ
+
𝛼
𝑡
)
⋅
ℎ
+
(
𝛽
ℓ
+
𝛽
𝑡
)
h
′
=(1+α
ℓ
	​

+α
t
	​

)⋅h+(β
ℓ
	​

+β
t
	​

)

# Define once:
self.layer_alpha = nn.Parameter(torch.zeros(self.n_layers))
self.layer_beta  = nn.Parameter(torch.zeros(self.n_layers))
self.task_alpha  = nn.ParameterDict({t: nn.Parameter(torch.zeros(1)) for t in task_names})
self.task_beta   = nn.ParameterDict({t: nn.Parameter(torch.zeros(1)) for t in task_names})

def condition_hidden(self, h, layer_idx, task):
    a = 1.0 + self.layer_alpha[layer_idx] + self.task_alpha[task]
    b = self.layer_beta[layer_idx] + self.task_beta[task]
    return h * a + b

B) Concat 2 scalars (adds 2 dims):

Concatenate 
[
𝑠
layerID
,
𝑠
taskID
]
[s
layerID
	​

,s
taskID
	​

] to each token hidden before the shared LM head.

self.layer_id = nn.Parameter(torch.zeros(self.n_layers, 1))    # scalar per layer
self.task_id  = nn.ParameterDict({t: nn.Parameter(torch.zeros(1)) for t in task_names})

def augment_hidden(self, h, layer_idx, task):
    B, T, D = h.shape
    lid = self.layer_id[layer_idx].expand(B, T, 1)
    tid = self.task_id[task].expand(B, T, 1)
    return torch.cat([h, lid, tid], dim=-1)  # -> (B,T,D+2)

2.4 Forward loop — compute per-layer losses for both tasks

Skeleton (inside forward, after each block):

layer_losses_tf = {}   # {f"layer_{i}": CE}
layer_losses_cp = {}   # {f"layer_{i}": CE}

for i, block in enumerate(self.blocks):
    x = block(x, in_span=in_span, span_id=span_id, is_prefix=is_prefix)

    # --- Teacher Forcing per-layer logits (shared head) ---
    if targets is not None:
        h = self.norm_out(x)
        if cfg.conditioning == "film":
            h = self.condition_hidden(h, i, "teacher_forcing")
            logits_i = self.shared_tf_head(h)
        else:  # "concat2"
            h_aug = self.augment_hidden(h, i, "teacher_forcing")
            logits_i = self.shared_tf_head(h_aug)

        # CE over tokens (mask/pad as you already do)
        ce_i = F.cross_entropy(logits_i.view(-1, logits_i.size(-1)),
                               targets.view(-1), ignore_index=PAD_ID)
        layer_losses_tf[f"layer_{i}"] = ce_i

    # --- Cocktail Party per-layer scores (contrastive span selection) ---
    if correct_idx is not None:
        h = self.norm_out(x)
        # compute context and span embeddings from h for this layer
        ctx_vec = extract_mask_query(h, tokens=x_tokens)          # (B, D)
        span_vecs = pool_spans(h, span_id)                        # (B, Nspans, D)
        # (Optional) condition h, or project ctx/span by a shared projection if desired
        # scores_i: dot(ctx, span) per candidate
        scores_i = torch.einsum('bd,bnd->bn', ctx_vec, span_vecs)
        ce_i = F.cross_entropy(scores_i, correct_idx)             # (B,)
        layer_losses_cp[f"layer_{i}"] = ce_i

# Final heads (unchanged)
logits_final = self.head(self.norm_out(x)) if targets is not None else None
scores_final = final_span_scores(self.norm_out(x), span_id) if correct_idx is not None else None

loss_tf = {"final_ce": ce_final_tf, "layer_ce": layer_losses_tf} if targets is not None else None
loss_cp = {"final_ce": ce_final_cp, "layer_ce": layer_losses_cp} if correct_idx is not None else None

return logits_final, scores_final, {"teacher_forcing": loss_tf, "cocktail_party": loss_cp}

Notes
• extract_mask_query, pool_spans, final_span_scores should mirror your existing final-layer cp logic—just reused per layer.
• We reuse one LM head across all layers; cp uses dot-product (no separate head per layer).
• Conditioning lets the shared LM head see layer/task identity with only scalars.

3) train_loop.py (or loss utils)
3.1 Apply uncertainty per (layer,task); keep reporting separate
def apply_layer_uncertainty_weighting(loss_dict, model, lambda_layers, lambda_kl):
    total = 0.0
    unc_deltas = {}   # for logging

    for task in ("teacher_forcing","cocktail_party"):
        if loss_dict.get(task) is None:
            continue

        Lfinal = loss_dict[task]["final_ce"]      # raw final CE (scalar)
        total += Lfinal                           # headline metric uses this raw CE

        # layer CE list/dict
        Llayers = loss_dict[task]["layer_ce"]     # {"layer_i": CE_i, ...}
        delta = 0.0

        for name, CE_i in Llayers.items():
            i = int(name.split('_')[-1])         # layer index
            s = model.layer_log_sigmas[i][task].clamp(-5, 5)   # s_{ℓ,t}
            weighted = 0.5 * torch.exp(-2*s) * CE_i + s
            total += lambda_layers[task] * weighted
            delta  += (0.5*torch.exp(-2*s) - 1.0) * CE_i + s

        # KL/L2 prior on s_{ℓ,t}
        s_all = [model.layer_log_sigmas[i][task] for i in range(model.n_layers)]
        kl = 0.5 * torch.stack([s**2 for s in s_all]).sum()
        total += lambda_kl * kl

        unc_deltas[task] = {"delta": delta.detach(), "kl": (lambda_kl*kl).detach()}

    return total, unc_deltas

3.2 Logging (don’t inflate headline loss)

Print final_raw CE per task as the task loss.

Log layer_mean_raw CE per task, and per-layer if desired.

Log uncertainty per task: delta, sigma_mean, and optional per-layer σ.

Example:

logs[f"loss/tf/final_raw"] = loss_dict["teacher_forcing"]["final_ce"].item()
logs[f"loss/tf/layer_mean_raw"] = torch.stack(
    list(loss_dict["teacher_forcing"]["layer_ce"].values())
).mean().item()

logs[f"uncertainty/tf/delta"] = unc_deltas["teacher_forcing"]["delta"].item()
logs[f"uncertainty/tf/sigma_mean"] = torch.stack(
    [model.layer_log_sigmas[i]["teacher_forcing"].exp()
     for i in range(model.n_layers)]
).mean().item()
# same for cp...

4) metrics.py / eval loop

Validation headline: val_loss/tf = final_raw_tf, val_loss/cp = final_raw_cp.

Optionally also log val_layer_mean_raw and uncertainty diagnostics; don’t mix into headline loss.

Options for the shared head + conditioning

conditioning = "film"

Zero extra dims; tiny param cost; fastest.

Head input stays (B,T,D).

conditioning = "concat2"

Appends two scalars (layer id, task id) → (B,T,D+2).

Single shared head over augmented features.

Both let one head serve all layers and both tasks (for LM). Cocktail party uses dot-product scoring (no extra heads).

Quick acceptance checklist

 layer_log_sigmas[ℓ][t] exists for all layers and both tasks; included in optimizer params.

 Forward loop computes layer_ce for tf and cp at every layer.

 Only one LM head (shared_tf_head) used at all layers; cp uses span dot-product.

 apply_layer_uncertainty_weighting uses s_{ℓ,t} (not task-level σ); adds KL on all s.

 Metrics: headline = final_raw CE per task; layer CE & uncertainty logged separately.

 Config can toggle conditioning mode and λ’s.
